src/tracksdata/graph/_test/test_graph_backends.py

```diff
@@ -1414,6 +1411,110 @@ def test_assign_track_ids(graph_backend: BaseGraph):
+    graph_backend.add_edge(B1, B4, {})
+    graph_backend.add_edge(B2, B3, {})
+    graph_backend.add_edge(B4, B5, {})
+
+    # Ensure track_id attribute exists after nodes were added
+    if DEFAULT_ATTR_KEYS.TRACK_ID not in graph_backend.node_attr_keys:
+        graph_backend.add_node_attr_key(DEFAULT_ATTR_KEYS.TRACK_ID, -1)
+
+    for seeds, expected in (
+        ([A1], [[A0, A1, A2, A3]]),
+        ([B2], [[B2, B3]]),
+        ([B1], [[B0, B1]]),
+        ([A1, B4], [[A0, A1, A2, A3], [B4, B5]]),
+    ):
+        # Ensure all nodes start unassigned across backends (avoid NULL/default discrepancies)
+        graph_backend.update_node_attrs(attrs={DEFAULT_ATTR_KEYS.TRACK_ID: -1})
+
+        tracks_graph = graph_backend.assign_track_ids(node_ids=seeds)
+        ids_df = graph_backend.node_attrs(attr_keys=[DEFAULT_ATTR_KEYS.NODE_ID, DEFAULT_ATTR_KEYS.TRACK_ID])
+        ids_map = dict(
+            zip(
+                ids_df[DEFAULT_ATTR_KEYS.NODE_ID].to_list(),
+                ids_df[DEFAULT_ATTR_KEYS.TRACK_ID].to_list(),
+                strict=True,
+            )
+        )
+        assigned = {}
+        for node_id, track_id in ids_map.items():
+            if track_id == -1:
+                continue
+            if track_id not in assigned:
+                assigned[track_id] = []
+            assigned[track_id].append(node_id)
+        assigned = {frozenset(group) for group in assigned.values()}
+        expected = {frozenset(group) for group in expected}
+        assert assigned == expected
+        assert isinstance(tracks_graph, rx.PyDiGraph)
+
+    # Check that re-assigning track IDs without reset works as expected
+    graph_backend.update_node_attrs(attrs={DEFAULT_ATTR_KEYS.TRACK_ID: -1})
+    tracks_graph = graph_backend.assign_track_ids()
+    ids_df = graph_backend.node_attrs(attr_keys=[DEFAULT_ATTR_KEYS.NODE_ID, DEFAULT_ATTR_KEYS.TRACK_ID])
+    assert tracks_graph.num_nodes() == 4
+
+    tracks_graph_reassign = graph_backend.assign_track_ids(node_ids=[A1, B4], reset=False)
+    ids_df_reassign = graph_backend.node_attrs(attr_keys=[DEFAULT_ATTR_KEYS.NODE_ID, DEFAULT_ATTR_KEYS.TRACK_ID])
+    assert tracks_graph_reassign.num_nodes() == 2
+    # Full reassignment should yield same result as previous partial reassignment
+    # df must be sorted by node_id for direct comparison
+    ids_df = ids_df.sort(DEFAULT_ATTR_KEYS.NODE_ID)
+    ids_df_reassign = ids_df_reassign.sort(DEFAULT_ATTR_KEYS.NODE_ID)
+    assert ids_df.equals(ids_df_reassign)
+
+    # Changing the topology
+    A4 = graph_backend.add_node({DEFAULT_ATTR_KEYS.T: 1, DEFAULT_ATTR_KEYS.TRACK_ID: -1})
+    graph_backend.remove_edge(A2, A3)
+    graph_backend.add_edge(A0, A4, {})
+    tracks_graph_reassign = graph_backend.assign_track_ids(node_ids=[A1, A4, B4], reset=False)
+    ids_df_reassign = graph_backend.node_attrs(attr_keys=[DEFAULT_ATTR_KEYS.NODE_ID, DEFAULT_ATTR_KEYS.TRACK_ID])
+    assert tracks_graph_reassign.num_nodes() == 3
```

Comment:
This test covers one of the edge cases I mentioned below.

Reply:
Right, I added a test and it was failing with the original impl.

Reply:
And now passing with the fixed one!